contracts/CourtPresaleActivate.sol

```diff
@@ -49,10 +59,93 @@ contract CourtPresaleActivate is IsContract, ApproveAndCallFallBack {
+    */
+    function contributeExternalToken(
+        address _token,
+        uint256 _amount,
+        uint256 _minTokens,
+        uint256 _minEth,
+        uint256 _deadline
+    )
+        external
+    {
+        require(_amount > 0, ERROR_ZERO_AMOUNT);
+
+        // move tokens to this contract
+        require(ERC20(_token).safeTransferFrom(msg.sender, address(this), _amount), ERROR_TOKEN_TRANSFER_FAILED);
+
+        // get the Uniswap exchange for the contribution token
+        address payable uniswapExchangeAddress = uniswapFactory.getExchange(_token);
+        require(uniswapExchangeAddress != address(0), ERROR_UNISWAP_UNAVAILABLE);
+        IUniswapExchange uniswapExchange = IUniswapExchange(uniswapExchangeAddress);
+
+        // swap tokens
+        address contributionTokenAddress = address(presale.contributionToken());
+        ERC20 token = ERC20(_token);
+        require(token.safeApprove(address(uniswapExchange), _amount), ERROR_TOKEN_APPROVAL_FAILED);
+        uint256 contributionTokenAmount = uniswapExchange.tokenToTokenSwapInput(_amount, _minTokens, _minEth, _deadline, contributionTokenAddress);
+
+        // buy in presale
+        _buyAndActivate(msg.sender, contributionTokenAmount, contributionTokenAddress);
+
+        // refund leftovers if any
+        _refund(token);
+        _refund(contributionTokenAddress);
+    }
+
+    /**
+    * @dev It will send the received ETH to Uniswap to get presale contribution tokens,
+    *      convert the obtained tokens in the presale instance, and activate the converted tokens in the
+    *      jurors registry instance of the Aragon Court.
+    * @param _minTokens Minimum amount of presale contribution tokens obtained in Uniswap
+    * @param _deadline Transaction deadline for Uniswap
+    */
+    function contributeEth(uint256 _minTokens, uint256 _deadline) external payable {
+        require(msg.value > 0, ERROR_ZERO_AMOUNT);
+
+        address contributionTokenAddress = address(presale.contributionToken());
+
+        // get the Uniswap exchange for the contribution token
+        address payable uniswapExchangeAddress = uniswapFactory.getExchange(contributionTokenAddress);
+        require(uniswapExchangeAddress != address(0), ERROR_UNISWAP_UNAVAILABLE);
+        IUniswapExchange uniswapExchange = IUniswapExchange(uniswapExchangeAddress);
+
+        // swap tokens
+        uint256 contributionTokenAmount = uniswapExchange.ethToTokenSwapInput.value(msg.value)(_minTokens, _deadline);
+
+        // buy in presale
+        _buyAndActivate(msg.sender, contributionTokenAmount, contributionTokenAddress);
+
+        // make sure there's no ETH left
+        uint256 ethBalance = address(this).balance;
+        if (ethBalance > 0) {
+            (bool result,) = msg.sender.call.value(ethBalance)("");
+            require(result, ERROR_ETH_REFUND);
+        }
+    }
+
+    function _buyAndActivate(address _from, uint256 _amount, address _token) internal {
```

Comment:
For internal functions, we might as well enforce type checking, so we could use `ERC20 _token`.